### ISSUE TITLE
check_swap: reports wrong swap size (~ factor 1024)

### ISSUE THREAD

**Issue by mrdsam** [monitoring-plugins/monitoring-plugins#1874]
Since Version 2.3.3 OR Version 2.3.2 (haven't testet this one) check_swap reports the wrong swap size by about factor 1024. Version 2.3.1 was still OK.

Both, plugin-output and performance-data are wrong.

My systems do have a 1024 MB swap partition, check_swap reports 1204223 MB. Here an example output, same system, 2 different check_swap binaries:

```
#
# New Version, wrong output
#
~$ /usr/local/libexec/nagios/check_swap --version
check_swap v (monitoring-plugins 2.3.3)

~$ /usr/local/libexec/nagios/check_swap -c 25% -w 50%
SWAP OK - 100% free (1204223MB out of 1204223MB) |swap=1262719336448B;631359668200;315679834100;0;1262719336448

#
# Older Version, right output
#
~$ ./check_swap --version
check_swap v (monitoring-plugins 2.3.1)

~$ ./check_swap -c 25% -w 50%
SWAP OK - 100% free (1023 MB out of 1023 MB) |swap=1023MB;511;255;0;1023
```
OS: FreeBSD 13.1-RELEASE-p6

**Comment by mrdsam** [comment]
Note: Because I can't tell whether this is a FreeBSD-only problem, I also opened a bug report at 
[bugs.freebsd.org](https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=271042) #271042
This bug report also links back here.

**Comment by sni** [comment]
the issue has been introduced by 4621427ba8cbfab4815e65d35a728ad51ad8c391

**Comment by RincewindsHat** [comment]
hm, feared that

**Comment by sni** [comment]
changing the swap_format to %*s %d %*d %d seems to fix it, but is has been there for ages.

```
freebsd13:/tmp/monitoring-plugins %>./plugins/check_swap -vvv -w 5% -c 5%
Command: /usr/sbin/swapinfo -k
Format: %*s %d %*d %d
total=0, free=0
SWAP OK - 100% free (4095MB out of 4095MB) |swap=4293918720B;214695935;214695935;0;4095
```
Then the only wrong value left is the total value at the end which should be bytes as well.

**Comment by RincewindsHat** [comment]
Interesting, I get a different swap format.

**Comment by RincewindsHat** [comment]
any chance, that the output of `swapinfo` changed?

**Comment by RincewindsHat** [comment]
we don't have error checking for all those `sscanf`s in there or am I wrong?

**Comment by RincewindsHat** [comment]
hm, works with `%*s %lu %*lu %lu` as the format

**Comment by sni** [comment]
no, we don't have tests, all we can do is run the test suite on those platforms. And no, i don't think the swapinfo output changed as it worked before the mentioned commit.
Also i haven't investigated why and with which scan format it worked before.

**Comment by Decstasy** [comment]
Same Problem here on a Debian 11:
```bash
$ '/usr/lib/nagios/plugins/check_swap' '-c' '25%' '-n' 'ok' '-w' '50%' -vvv; cat /proc/meminfo; free -k
Reading PROC_MEMINFO at /proc/meminfo
Got Cached with 568564
Got Total with 3997692
Got Free with 16580
SWAP CRITICAL - 15% free (571MB out of 3903MB) |swap=598736896B;2046296050;1023148025;0;4092592128
MemTotal:       41079844 kB
MemFree:         1468664 kB
MemAvailable:    9128696 kB
Buffers:         1086100 kB
Cached:          4259316 kB
SwapCached:       568564 kB
Active:         23968984 kB
Inactive:       12150088 kB
Active(anon):   20801568 kB
Inactive(anon): 10037232 kB
Active(file):    3167416 kB
Inactive(file):  2112856 kB
Unevictable:           0 kB
Mlocked:               0 kB
SwapTotal:       3997692 kB
SwapFree:          16580 kB
Dirty:              2872 kB
Writeback:             0 kB
AnonPages:      30045712 kB
Mapped:           355460 kB
Shmem:             65140 kB
KReclaimable:    2910912 kB
Slab:            3186384 kB
SReclaimable:    2910912 kB
SUnreclaim:       275472 kB
KernelStack:       34224 kB
PageTables:       102820 kB
NFS_Unstable:          0 kB
Bounce:                0 kB
WritebackTmp:          0 kB
CommitLimit:    24537612 kB
Committed_AS:   44884492 kB
VmallocTotal:   34359738367 kB
VmallocUsed:       97096 kB
VmallocChunk:          0 kB
Percpu:             6960 kB
HardwareCorrupted:     0 kB
AnonHugePages:  25430016 kB
ShmemHugePages:        0 kB
ShmemPmdMapped:        0 kB
FileHugePages:         0 kB
FilePmdMapped:         0 kB
HugePages_Total:       0
HugePages_Free:        0
HugePages_Rsvd:        0
HugePages_Surp:        0
Hugepagesize:       2048 kB
Hugetlb:               0 kB
DirectMap4k:     2793408 kB
DirectMap2M:    37052416 kB
DirectMap1G:     4194304 kB
               total        used        free      shared  buff/cache   available
Mem:        41079844    31355120     1468396       65140     8256328     9128428
Swap:        3997692     3981112       16580

```

**Comment by RincewindsHat** [comment]
@Decstasy, is this really the same problem? It looks like you actually have less than 25% free swap

**Comment by waja** [comment]
> Same Problem here on a Debian 11:
> 
> ```bash
> 
> $ '/usr/lib/nagios/plugins/check_swap' '-c' '25%' '-n' 'ok' '-w' '50%' -vvv; cat /proc/meminfo; free -k
> 
> [...]
> 
> ```

Do you use the Debian package in bullseye or did you compile it on your own?

**Comment by Decstasy** [comment]
@waja Tested both, same result.

@RincewindsHat I want to point out the following `Got Free with 16580` then it printed `15% free (571MB` which is not correct there are just 16580 kb free which are around 16MB. So where does 571MB come from?
If this Issue is not about that, I'm sorry.

**Comment by RincewindsHat** [comment]
@Decstasy I think it is a different problem, you are on a linux, so this exact problem should not happen to you.
Could you open another issue here, so we have a proper separation of concerns?

**Comment by RincewindsHat** [comment]
ah, found it. @Decstasy Cache (`SwapCached`) is counted as free, since it is not actually in use anymore (although the pages still exist in swap memory if they need to be swapped out again)).
This explains the number and is arguably correct. Would be better if it would actually display that more obviously

**Comment by RincewindsHat** [comment]
seems like this can be closed

**Comment by RincewindsHat** [comment]
@mrdsam could you try #1876 ?

**Comment by Napsty** [comment]
@RincewindsHat Are you sure this is actually solved? Compiling monitoring-plugins from the current master shows:

```
$ ./check_swap --version
check_swap v2.3.5.85.g1a82 (monitoring-plugins 2.3git)

$ ./check_swap -w 10% -c 5%
SWAP CRITICAL - 2% free (136MB out of 7627MB) |swap=142606336B;799748910;399874455;0;7997489152
```

The system (Debian 10) has no swap left...

![image](https://github.com/monitoring-plugins/monitoring-plugins/assets/3702203/a8c980a4-d01e-4d91-969b-8bca848676ba)

Comparing with the check_swap from the monitoring-plugins Debian package:

```
$ /usr/lib/nagios/plugins/check_swap --version
check_swap v2.2 (monitoring-plugins 2.2)

$ /usr/lib/nagios/plugins/check_swap -w 10% -c 5%
SWAP CRITICAL - 0% free (0 MB out of 7627 MB) |swap=0MB;762;381;0;7627
```

And also compared to check_swap from nagios-plugins (git master):

```
$ /tmp/nagios-plugins/plugins/check_swap -w 10% -c 5%
SWAP CRITICAL - 0% free (0 MB out of 7627 MB) |swap=0MB;762;381;0;7627
```

I was working on a PR of check_swap (for something else) when I noticed the erroneous output.

**Comment by RincewindsHat** [comment]
@Napsty weird, works for me spontaneously on debian 13 (trixie). I'll take a look, thank you

**Comment by RincewindsHat** [comment]
@Napsty could you paste the output of the `master` branch with `-vvv`?

**Comment by Napsty** [comment]
Sure, here we go:

```
$ ./check_swap --version
check_swap v2.3.5.85.g1a82 (monitoring-plugins 2.3git)

$ ./check_swap -w 10% -c 5% -vvv
Reading PROC_MEMINFO at /proc/meminfo
Got Cached with 132500
Got Total with 7811068
Got Free with 0
SWAP CRITICAL - 2% free (129MB out of 7627MB) |swap=135266304B;799748910;399874455;0;7997489152
```

Verbose 3 output of check_swap from deb package:

```
$ /usr/lib/nagios/plugins/check_swap --version
check_swap v2.2 (monitoring-plugins 2.2)

$ /usr/lib/nagios/plugins/check_swap -w 10% -c 5% -vvv
Reading PROC_MEMINFO at /proc/meminfo
Got Total with 7811068.000000
Got Free with 0.000000
SWAP CRITICAL - 0% free (0 MB out of 7627 MB) |swap=0MB;762;381;0;7627
```

Outpout from `cat /proc/meminfo`:

```
MemTotal:       32877388 kB
MemFree:         1282964 kB
MemAvailable:    4883984 kB
Buffers:         1719656 kB
Cached:          2616692 kB
SwapCached:       132500 kB
Active:         10726912 kB
Inactive:       19369664 kB
Active(anon):    8126652 kB
Inactive(anon): 18463188 kB
Active(file):    2600260 kB
Inactive(file):   906476 kB
Unevictable:      133620 kB
Mlocked:          133620 kB
SwapTotal:       7811068 kB
SwapFree:              0 kB
Dirty:              1060 kB
Writeback:           320 kB
AnonPages:      25772688 kB
Mapped:          1110496 kB
Shmem:           1020716 kB
KReclaimable:     566544 kB
Slab:             925580 kB
SReclaimable:     566544 kB
SUnreclaim:       359036 kB
KernelStack:       43584 kB
PageTables:       200824 kB
NFS_Unstable:          0 kB
Bounce:                0 kB
WritebackTmp:          0 kB
CommitLimit:    24249760 kB
Committed_AS:   57061656 kB
VmallocTotal:   34359738367 kB
VmallocUsed:       90724 kB
VmallocChunk:          0 kB
Percpu:            28800 kB
HardwareCorrupted:     0 kB
AnonHugePages:   7458816 kB
ShmemHugePages:        0 kB
ShmemPmdMapped:        0 kB
FileHugePages:         0 kB
FilePmdMapped:         0 kB
HugePages_Total:       0
HugePages_Free:        0
HugePages_Rsvd:        0
HugePages_Surp:        0
Hugepagesize:       2048 kB
Hugetlb:               0 kB
DirectMap4k:    13635928 kB
DirectMap2M:    19847168 kB
DirectMap1G:     1048576 kB
```

**Comment by Napsty** [comment]
The problem resides in the calculation of free swap, used as `dskfree_mb` inside the plugin. It assumes that `SwapCached` from `/proc/meminfo` is counted as free swap - but it isn't. 

Created PR #1972 which should fix the wrong free output. 

With the fix applied:

```
$ ./plugins/check_swap -w 10% -c 5%
SWAP CRITICAL - 0% free (0MB out of 7627MB) |swap=0B;799748910;399874455;0;7997489152

$ ./plugins/check_swap -w 10% -c 5% -vvv
Reading PROC_MEMINFO at /proc/meminfo
Got Cached with 132816
Got Total with 7811068
Got Free with 0
SWAP CRITICAL - 0% free (0MB out of 7627MB) |swap=0B;799748910;399874455;0;7997489152
```

Not sure what to do with the "Cached" value though. For now I just placed it into a separate variable `dskcached_mb`. Not sure if this is of use inside the perfdata or if it even makes sense to parse this.

**Comment by RincewindsHat** [comment]
from `man 5 proc`

```
SwapCached %lu
  Memory  that  once  was  swapped  out, is swapped back in but still also is in the swap file.  (If memory pressure is high, these pages don't need to be swapped out again because they are already in the swap file.
  This saves I/O.)
```

I would count that as free.

**Comment by Napsty** [comment]
It's definitely a case of interpretation, but I personally understand this as _swap used_ and definitely not as swap free:

> because they are already in the swap file

Other observations:

- "What swap cache is for files on disk, SwapCached is for ANON pages on a swap device." https://unix.stackexchange.com/a/578549/321061
- "the field SwapCached is unrelated; it has nothing to do with any disk cache or slab cache. It's how much of the swap is also present in the clean memory pages." https://unix.stackexchange.com/a/334525/321061
- "SwapCached: Memory that is present within main memory, but also in the swapfile." https://access.redhat.com/solutions/406773
- "SwapCached: The amount of swap, in kilobytes, used as cache memory." https://support.hpe.com/hpesc/public/docDisplay?docId=c02015507&docLocale=en_US

So basically everyone counts "SwapCached" as swap used, not as swap free. 

We could add this as a separate information in output and performance data, or do something else with that value, but definitely not count it as "free" (that's what the Kernel is supposed to do on its own to come up with the SwapFree value).

Or as a different idea, offer something like check_mem (https://github.com/justintime/nagios-plugins/tree/master/check_mem): Add an option to let the user decide whether the Cache should be counted as used or free. But this would mean a major (compatibility) step away from check_swap on nagios-plugins and could interfere with large monitoring environments with mixed servers and distributions.

I'd keep it simple though.

**Comment by RincewindsHat** [comment]
hm, I see, there are different interpretations of "free". Mine would have been "space which can be used for swapping" and to my understanding the cached space is available for that, although it is not "unused".

**Comment by Napsty** [comment]
Agree with you, I think there are different ways to understand what "free" means. 
To me, free means the actual free space of swap, not including any used or cached (or temporary) pages.
Now how do we solve this? What should the plugin use as "free"? 

By looking at other tools (htop, top, check_swap from nagios-plugins) they all don't account SwapCached as "free". To me personally, I would go the same path as the other tools/scripts. Otherwise this leads to confusion. 
But it's up to you guys (monitoring-plugins maintainers) to decide on that.

**Comment by RincewindsHat** [comment]
@sni @waja opinions on that?

**Comment by RincewindsHat** [comment]
As discussed in #1972 the default behavior stays as it is now.
